### PR TITLE
add query controller and models

### PR DESCRIPTION
**Pull request by BrandonArp**

### CODE REVIEW

File: app/models/internal/impl/DefaultMetricsQueryResult.java [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+import net.sf.oval.constraint.Min;
+import net.sf.oval.constraint.NotEmpty;
+import net.sf.oval.constraint.NotNull;
+import org.joda.time.DateTime;
+
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Model class to represent a metrics query response from KairosDB.
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultMetricsQueryResult.java]: If this is a KairosDb model it should probably not be in the `internal` package.

File: app/models/internal/impl/DefaultAlertTrigger.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        _endTime = builder._endTime;
+        _args = builder._args;
+        _message = builder._message;
+        _groupBy = builder._groupBy;
+    }
+
+    private final DateTime _time;
+    private final Optional<DateTime> _endTime;
+    private final String _message;
+    private final ImmutableMap<String, String> _groupBy;
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultAlertTrigger.java]: What is the group-by on the trigger?

File: app/models/internal/AlertTrigger.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * The arguments (tags) for the trigger.
+     *
+     * @return the args
+     */
+    ImmutableMap<String, String> getArgs();
+
+    /**
+     * The group by (tags) for the trigger. These determine uniqueness for a trigger.
```
Comment (vjkoskela) [inline app/models/internal/AlertTrigger.java]: Can you elaborate on this? (dup comment below on impl)

File: app/global/MainModule.java [hunk 4]
```diff
@@ -81,6 +82,9 @@ protected void configure() {
         bind(OrganizationProvider.class)
                 .toProvider(ConfigTypedProvider.provider("organizationProvider.type"))
                 .in(Scopes.SINGLETON);
+        bind(QueryExecutor.class)
+                .toProvider(ConfigTypedProvider.provider("query.executor.type"))
```
[the remark this answers is not in the document]
Comment (BrandonArp) [reply app/global/MainModule.java]: Done! Added a binding string -> executor.

**Comment by BrandonArp** [comment]
Ready for the next round of fun

File: conf/portal.routes [hunk 5]
```diff
@@ -46,6 +46,9 @@ GET        /v1/reports/:id                      controllers.ReportController.get
 DELETE     /v1/reports/:id                      controllers.ReportController.delete(id: java.util.UUID)
 PUT        /v1/reports                          controllers.ReportController.addOrUpdate
 
+# Metrics Queries
+POST       /v1/metrics/query                    controllers.MetricsController.query()
```
Comment (vjkoskela) [inline conf/portal.routes]: Should we remove this for now until we iterate on this design some more?

**Pull request merged**